File: src/main/java/nl/hand/made/vending/machine/model/mapper/ProductMapper.java
package nl.hand.made.vending.machine.model.mapper;

import nl.hand.made.vending.machine.exception.ValidationException;
import nl.hand.made.vending.machine.model.request.ProductData;
import nl.hand.made.vending.machine.model.response.ProductResponse;
import nl.hand.made.vending.machine.repository.model.ProductEntity;
import nl.hand.made.vending.machine.repository.model.UserEntity;
import nl.hand.made.vending.machine.util.AmountUtil;
import org.apache.commons.lang3.StringUtils;

public class ProductMapper {

    public static ProductResponse mapToResponseModel(ProductEntity productEntity) {
        ProductResponse productResponse = new ProductResponse();
        productResponse.setProductName(productEntity.getProductName());
        productResponse.setAmountAvailable(productEntity.getAmountAvailable());
        productResponse.setCost(productEntity.getCost());
        productResponse.setSeller(productEntity.getSeller().getUsername());
        return productResponse;
    }

    public static ProductEntity mapToEntityModel(ProductData productData, UserEntity sellerEntity) {
        ProductEntity productEntity = new ProductEntity();
        productEntity.setProductName(productData.getProductName());
        productEntity.setCost(productData.getCost());
        productEntity.setAmountAvailable(productData.getAmountAvailable());
        productEntity.setSeller(sellerEntity);
        return productEntity;
    }

    public static ProductEntity patch(ProductEntity productEntity, ProductData productData) {
        if (StringUtils.isNotBlank(productData.getProductName())) {
            productEntity.setProductName(productData.getProductName());
        }

        if (productData.getAmountAvailable() != null) {
            if (productData.getAmountAvailable() < 0) {
                throw new ValidationException("product cost cannot be negative");
            }
            productEntity.setAmountAvailable(productData.getAmountAvailable());
        }

        if (productData.getCost() != null) {
            AmountUtil.validateProductCost(productData.getCost());
            productEntity.setCost(productData.getCost());
        }

        return productEntity;
    }
}

File: src/main/kotlin/nl/hand/made/vending/machine/model/mapper/ProductMapper.kt
package nl.hand.made.vending.machine.model.mapper

import nl.hand.made.vending.machine.exception.ValidationException
import nl.hand.made.vending.machine.model.request.ProductCreate
import nl.hand.made.vending.machine.model.request.ProductPatch
import nl.hand.made.vending.machine.model.response.ProductResponse
import nl.hand.made.vending.machine.repository.model.ProductEntity
import nl.hand.made.vending.machine.repository.model.UserEntity
import nl.hand.made.vending.machine.util.AmountUtil

class ProductMapper {

    companion object {
        fun mapToResponseModel(productEntity: ProductEntity) =
            ProductResponse(
                productName = productEntity.productName,
                amountAvailable = productEntity.amountAvailable,
                cost = productEntity.cost,
                seller = productEntity.seller.username
            )

        fun mapToEntityModel(productData: ProductCreate, sellerEntity: UserEntity) =
            ProductEntity(
                productName = productData.productName,
                cost = productData.cost,
                amountAvailable = productData.amountAvailable,
                seller = sellerEntity
            )

        fun patch(productEntity: ProductEntity, productData: ProductPatch): ProductEntity {
            productData.productName?.run {
                this.isNotBlank() || throw ValidationException("product name cannot be empty")
                productEntity.productName = this
            }

            productData.amountAvailable?.run {
                this >= 0 || throw ValidationException("amount available cannot be negative")
                productEntity.amountAvailable = this
            }

            productData.cost?.run {
                AmountUtil.validateProductCost(this)
                productEntity.cost = this
            }
            return productEntity;
        }
    }
}